src/basic/hostname-util.c

```diff
@@ -221,6 +221,39 @@ int sethostname_idempotent(const char *s) {
         return 1;
 }
 
+int shorten_overlong(const char *s, char **ret) {
+        char *h, *p;
+
+        /* Shorten an overlong name to HOST_NAME_MAX or to the first dot,
+         * whatever comes earlier. */
+
+        assert(s);
+
+        h = strdup(s);
+        if (!h)
+                return -ENOMEM;
+
+        if (strlen(h) <= HOST_NAME_MAX) {
+                *ret = h;
+                return 0;
+        }
+
+        for (p = h; *p; p++) {
+                if (*p == '.') {
+                        *p = 0;
+                        break;
+                }
+        }
+
+        strshorten(h, HOST_NAME_MAX);
```

Comment:
hmm, this should not be necessary, the first label should not be longer than 64 chars anyway...

Reply:
Not sure I got your point, input could still be a single overlong label like at https://github.com/systemd/systemd/pull/7616/files#diff-cedcecfd5de09ead4c57bf267ed1e705R223.

Reply:
are you sure it could be? iirc dhcp encodes the fqdn in DNS-style binary encoding, and that doesn't cover labels longer than 64 chars, i.e. the encoding already makes that impossible...

Reply:
(but anyway, i don't really care about this one, i just don't hink it will ever have any effect)

Reply:
Now I see. No, I think that this specific DHCP option is just a single-byte length plus a sequence of characters: https://tools.ietf.org/html/rfc2132#section-3.17

Reply:
ah, ok, then that makes sense.